ndcube/utils/wcs.py

```diff
@@ -300,3 +158,270 @@ def _pixel_keep(wcs_object):
+    """
+    Gets the world axis physical types corresponding to a pixel axis.
+
+    Parameters
+    ----------
+    pixel_axis: `int`
+        The pixel axis number(s) for which the world axis numbers are desired.
+
+    wcs: `astropy.wcs.BaseLowLevelWCS`
+        The WCS object defining the relationship between pixel and world axes.
+
+    Returns
+    -------
+    physical_types: `numpy.ndarray` of `str`
+        The physical types corresponding to the pixel axis.
+    """
+    return np.array(wcs.world_axis_physical_types)[wcs.axis_correlation_matrix[:, pixel_axis]]
+
+
+def physical_type_to_pixel_axes(physical_type, wcs):
+    """
+    Gets the pixel axis indices corresponding to a world axis physical type.
+
+    Parameters
+    ----------
+    physical_type: `int`
+        The pixel axis number(s) for which the world axis numbers are desired.
+
+    wcs: `astropy.wcs.BaseLowLevelWCS`
+        The WCS object defining the relationship between pixel and world axes.
+
+    Returns
+    -------
+    pixel_axes: `numpy.ndarray`
+        The pixel axis indices corresponding to the physical type.
+    """
+    world_axis = physical_type_to_world_axis(physical_type, wcs.world_axis_physical_types)
+    return world_axis_to_pixel_axes(world_axis, wcs.axis_correlation_matrix)
+
+
+def physical_type_to_world_axis(physical_type, world_axis_physical_types):
+    """
+    Returns world axis index of a physical type based on WCS world_axis_physical_types.
+
+    Input can be a substring of a physical type, so long as it is unique.
+
+    Parameters
+    ----------
+    physical_type: `str`
+        The physical type or a substring unique to a physical type.
+
+    world_axis_physical_types: sequence of `str`
+        All available physical types.  Ordering must be same as
+        `astropy.wcs.BaseLowLevelWCS.world_axis_physical_types`
+
+    Returns
+    -------
+    world_axis: `numbers.Integral`
+        The world axis index of the physical type.
+    """
+    # Find world axis index described by physical type.
+    widx = np.where(world_axis_physical_types == physical_type)[0]
+    # If physical type does not correspond to entry in world_axis_physical_types,
+    # check if it is a substring of any physical types.
+    if len(widx) == 0:
+        widx = [physical_type in world_axis_physical_type
+                for world_axis_physical_type in world_axis_physical_types]
+        widx = np.arange(len(world_axis_physical_types))[widx]
+    if len(widx) != 1:
+        raise ValueError(
+                "Input does not uniquely correspond to a physical type."
+                f" Expected unique substring of one of {world_axis_physical_types}."
+                f"  Got: {physical_type}")
+    # Return axes with duplicates removed.
+    return widx[0]
+
+
+def get_dependent_pixel_axes(pixel_axis, axis_correlation_matrix):
+    """
+    Given a WCS pixel axis index, return indices of dependent WCS pixel axes.
```

Comment:
Does the return include the input?

```suggestion
    Given a WCS pixel axis index, return indices of all dependent WCS pixel axes.
```

This is a little weird to reason about as well. Because pixel axes aren't dependant on other pixel axes, the pixel axes are always separable (they just might have more than one world axis associated with them). So I think this description really needs rewriting to make it clear what's happening.

Reply:
That's fair enough.  I'll change it to the following:
```
"""
Return indices of all pixel axes associated with the world axes that correspond to the input pixel axis.

For example, say the input pixel axis is 0 and it is associated with two world axes corresponding to longitude and latitude.
Let's also say that pixel axis 1 is also associated with longitude and latitude.
Thus, this function would return pixel axes 0 and 1.
On the other hand let's say pixel axis 2 is associated with only one world axis, say wavelength, which does not depend on any other pixel axis (i.e. it is independent).
In that case this function would only return pixel axis 2.
"""
```